JuliaSyntax/src/porcelain/syntax_graph.jl

```diff
@@ -857,3 +875,488 @@ function _map_green_to_ast(parent::Kind, cs::SyntaxList; eq_to_kw=false)
+        # (let (block (= x y)) _...) => (let (= x y) _...)
+        # (let (block (:: x y)) _...) => (let (:: x y) _...)
+        # (struct _ (block (doc "foo" field1) (doc "bar" field2))) =>
+        # (struct _ (block "foo" field1 "bar" field2))
+        if kind(parent) === K"let" && parent_i === 1 && n_cs === 1
+            out = _green_to_est(st, 1, cs[1])
+            return kind(out) in KSet"Identifier = ::" ? out :
+                mknode(st, tree_ids(out))
+        elseif kind(parent) === K"struct" && parent_i === 3
+            cs_tmp = SyntaxList(cs)
+            for c in cs
+                kind(c) === K"doc" ?
+                    append!(cs_tmp, preprocessed_green_children(c)) :
+                    push!(cs_tmp, c)
+            end
+            cs = cs_tmp
+        end
+    elseif (k === K"local" || k === K"global") && n_cs === 1
+        # (local (const _)) => (const (local _))
+        # (local (tuple a b c)) => (local a b c)
+        if kind(cs[1]) === K"const"
+            ret_c1_cs = _map_green_to_est(st, preprocessed_green_children(cs[1]))
+            ret_cs = tree_ids(mknode(st, ret_c1_cs))
+            return mknode(cs[1], ret_cs)
+        elseif kind(cs[1]) === K"tuple"
+            cs = preprocessed_green_children(cs[1])
+        end
+    elseif k === K"return" && n_cs === 0
+        push!(cs, coreref("nothing"))
+    elseif k === K"juxtapose"
+        ret_k = K"call"
+        pushfirst!(cs, symleaf("*"))
+    elseif k === K"struct"
+        is_mutable = valleaf(has_flags(st, MUTABLE_FLAG))
+        pushfirst!(cs, is_mutable)
+    elseif k === K"importpath"
+        ret_k = K"."
+        for i in eachindex(cs)
+            if kind(cs[i]) === K"inert"
+                inner_cs = preprocessed_green_children(cs[i])
+                length(inner_cs) === 1 && (cs[i] = only(inner_cs))
+            end
+        end
+    elseif k === K"wrapper" # parse errors only
+        ret_k = K"block"
+    elseif k === K"parameters"
+        kw_in_params = kind(parent) === K"parameters" && parent_i === 1 ?
+            kw_in_params : !(kind(parent) in KSet"vect curly braces ref")
+    elseif k === K"="
+        p_k = kind(parent)
+        because_params = p_k === K"parameters" && parent_i >= 1 && kw_in_params
+        because_call = parent_i > 1 && (p_k == K"ref" ||
+            p_k in KSet"call dotcall" && is_prefix_call(parent))
+        ret_k = because_params || because_call ? K"kw" : K"="
+    end
+
+    # Recurse on `cs`.  If no children change, just return `st`.
+    ret_cs = _map_green_to_est(st, cs; kw_in_params)
+    return ret_cs.ids == children(st).ids && ret_k == kind(st) ?
+        st : setattr!(mknode(st, ret_cs), :kind, ret_k)
+end
+
+function _map_green_to_est(parent::SyntaxTree, cs::SyntaxList;
+                           kw_in_params=false, undef_parent=false)
+    ret_cs = SyntaxList(cs.graph)
+    for (i, c) in enumerate(cs)
+        undef_parent && (i = 0)
+        new_c = _green_to_est(parent, i, c; kw_in_params)
```

Comment:
```suggestion
        new_c = _green_to_est(parent, undef_parent ? 0 : i, c; kw_in_params)
```

Modifying an induction variable in-place, even if it is harmless / temporary can be confusing imo